### ISSUE TITLE
How do you get solution vector?

### ISSUE THREAD

**Issue by lonelykid** [kailaix/ADCME.jl#2]
Greetings!

I have questions about how you treat the solution vector. In the simple example shown here, you just used `u = B \ f`. However, in the paper, it appears to me the complete solution vector is obtained from a FEM solution, kept unchanged and used to calibrate ![](http://latex.codecogs.com/gif.latex?\\theta) in the training process. Theoretically, you could also use something like `u = B \ f`.  And using previous ![](http://latex.codecogs.com/gif.latex?u) is impractical for experimental data as you cannot get the full solution vector. Are there any issues combining autograd and solving a large linear system?

**Comment by kailaix** [comment]
@lonelykid In the paper, the system was assembled using FEM but the resultant equation P(u, Mθ(u, x)) = F(u, x, p) was **nonlinear** in u in the multiscale model. To solve the nonlinear equation, we increased p gradually to the true value (incremental load method) and applied Newton-Raphson methods. This may be difficult to implement directly in AD. 
Since we are considering a 2D problem, the history of u is measurable in a laboratory. In 3D we are only able to measure parts of data (e.g., on the surface). In this case, we can use custom operators (see [More Documentation](https://github.com/kailaix/ADCME.jl#more-documentation)), apply the previously mentioned to obtain the solution vector and compare the measurable parts with observed data.

**Comment by lonelykid** [comment]
Thanks for your timely reply. This clarified what I had in mind. It may be hard for autograd when sophisticated solvers like NR is involved.  I'm glad you had no issue with autograd for just 'inverting' a large matrix.

I'm not sure how you would do it in the 3D case. Your custom-op example seems clarified my first question not this one.  For a 3D nonlinear problem, you still need the full solution vector to construct your Pint and Fext, you cannot construct it only on the surface. I guess this is something for further study? If you can combine NR and autograd, this is not a problem, though.

**Comment by kailaix** [comment]
The idea behind ADCME is that you assume that you parametrize your the constitutive law with fθ. Since you now know the constitutive law and boundary conditions, you can apply the NR method and obtain u full solution vector in "Tensor" form (not possible in direct AD, but it can be achieved with custom operators). Now let's assume you observe u_{obs} on the surface, and the corresponding numerical solution is u[I_{obs}], I_{obs} are the DOFs of u corresponding to the surface. You can have a loss function like ||u[I_{obs}]-u_{obs}||^2. By minimizing the loss function you can estimate θ. 
BTW, we need to be careful when "inverting" large matrices in AD; AD is memory consuming and does not have specialized linear solvers in general. 
Hope that answers your question :)

**Comment by lonelykid** [comment]
You mean you need to implement NR method in custom_op?  I'm not sure if I understand you correctly.  But you also need to implement backward for NR. I don't know whether this is doable. 
BTW, I agree doing a FEM run first is acceptable. You mentioned about constructing a coarse-grained model.
Thanks!

**Comment by kailaix** [comment]
Yes, it's actually doable. You may not want to implement backward for NR but for the nonlinear operator that NR is trying to solve. Suppose you have a nonlinear operator x->y, F(x, y) = 0, for forward you use NR to compute y given x. For backward, you need to implement d(loss)/dx = d(loss)/dy * dy/dx. See [here](https://github.com/kailaix/ADCME.jl/blob/master/examples/md/four_types.ipynb).

**Comment by lonelykid** [comment]
Thanks for the example. Looks indeed doable but challenging to implement for NR in FEM.

**Comment by kailaix** [comment]
Yes, it's indeed challenging. Let me know if you have thoughts on that :)
